core/src/main/scala/BinaryTree.scala

```diff
@@ -0,0 +1,251 @@
+/**
+ * Created by Nicolas A Perez (@anicolaspp) on 1/29/16.
+ */
+
+package dogs
+
+import Predef._
+import dogs.Order.{GT, EQ, LT}
+import scala.annotation.tailrec
+import scala.math
+
+
+
+abstract class BinaryTree[A] {
+
+  def value: Option[A]
+
+  val left: BinaryTree[A]
+  val right: BinaryTree[A]
+
+  private [dogs] val size: Int
+  private[dogs] val height: Int
+  private [dogs] def balanceFactor: Int
+  private[dogs] def update(): BinaryTree[A]
+
+  def isEmpty: Boolean
+
+  def toLst(): List[A] = this match {
+    case BTNil()                      =>  El[A]
+    case Branch(Some(a), l, r, _, _)  =>  l.toLst ::: (a :: r.toLst)
+  }
+
+  def inorder() = toLst()
+
+  def preorder(): List[A] = this match {
+    case BTNil()                      =>  El[A]
+    case Branch(Some(a), l, r, _, _)  =>  a :: l.preorder ::: r.preorder
+  }
+
+  def posorder(): List[A] = this match {
+    case BTNil()                      =>  El[A]
+    case Branch(Some(a), l, r, _, _)  =>  l.posorder() ::: r.posorder() ::: Nel(a, El[A])
+  }
+
+  def min(): Option[A] =  {
+    @tailrec def loop(sub: BinaryTree[A], x: Option[A]): Option[A] = sub match {
+      case BTNil()              =>  x
+      case Branch(a, l, _, _, _)      =>  loop(l, a)
+    }
+
+    this match {
+      case BTNil()                    => None()
+      case Branch(a, l, _, _, _)      =>  loop(l, a)
+    }
+  }
+
+  def max(): Option[A] = {
+    @tailrec def loop(sub: BinaryTree[A], x: Option[A]): Option[A] = sub match {
+      case BTNil()                    =>  x
+      case Branch(a, _, r, _, _)      =>  loop(r, a)
+    }
+
+    this match {
+      case BTNil()                    =>  None()
+      case Branch(a, _, r, _, _)      =>  loop(r, a)
+    }
+  }
+
+  def contains(x: A)(implicit order: Order[A]): Boolean = this match {
+    case BTNil()                      =>  false
+
+    case Branch(Some(a), l, r, _, _)  => order.compare(x, a) match {
+      case LT                             =>  l.contains(x)
+      case EQ                             =>  true
+      case GT                             =>  r.contains(x)
+    }
+  }
+
+  def add(x: A)(implicit order: Order[A]): Branch[A] = {
+    def insert(x: A, order: Order[A])  = this match {
+      case BTNil()                    =>  Branch(Some(x), BinaryTree.empty, BinaryTree.empty)
+      case Branch(Some(a), l, r, _, _)=>  order.compare(x, a) match {
+        case LT                     =>  Branch(Some(a), l.add(x)(order), r)
+        case EQ                     =>  this.asInstanceOf[Branch[A]]
```

Comment:
can you tell I really am uncomfortable with asInstanceeOf yet?  :)

you can change

    case Branch(Some(a)...

to 
    
    case b @ Branch(Some(a)....

then instead of this.asInstanceOf, you can return b